File: src/main/java/26.remove-duplicates-from-sorted-array.java
class Q26 {
  public int removeDuplicates(int[] nums) {
    if (nums.length < 2) {
      return nums.length;
    }
    int k = 1;
    for (int i = 1; i < nums.length; i++) {
      if (nums[i] != nums[i - 1]) {
        nums[k++] = nums[i];
      }
    }
    return k;
  }
}

File: src/main/scala/26.remove-duplicates-from-sorted-array.scala
/**
 * 思路:
 * 利用双指针, 当p != q 时把不相等的部分挪到数组的前半部分
 */
object Q26Scala {
  def removeDuplicates(nums: Array[Int]): Int =
    if (nums.length == 0) 0
    else {
      @scala.annotation.tailrec
      def loop(p: Int, q: Int): Int =
        if (q == nums.length) p + 1
        else if (nums(p) != nums(q)) {
          nums(p + 1) = nums(q)
          loop(p + 1, q)
        } else loop(p, q + 1)
      loop(0, 1)
    }
}
